key_derivation.go

```diff
@@ -5,56 +5,28 @@ import (
 	"encoding/binary"
 )
 
-// All of these key derivation functions are AES-CM specific
-// in the future we have multiple implementations of each of these functions
-
-func generateSessionKey(label byte, masterKey, masterSalt []byte) ([]byte, error) {
+func aesCmKeyDerivation(label byte, masterKey, masterSalt []byte, outLen int) ([]byte, error) {
 	// https://tools.ietf.org/html/rfc3711#appendix-B.3
 	// The input block for AES-CM is generated by exclusive-oring the master salt with the
 	// concatenation of the encryption key label 0x00 with (index DIV kdr),
 	// - index is 'rollover count' and DIV is 'divided by'
-	sessionKey := make([]byte, len(masterSalt))
-	copy(sessionKey, masterSalt)
-
-	labelAndIndexOverKdr := []byte{label, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
-	for i, j := len(labelAndIndexOverKdr)-1, len(sessionKey)-1; i >= 0; i, j = i-1, j-1 {
-		sessionKey[j] = sessionKey[j] ^ labelAndIndexOverKdr[i]
-	}
-
-	// then padding on the right with two null octets (which implements the multiply-by-2^16 operation, see Section 4.3.3).
-	sessionKey = append(sessionKey, []byte{0x00, 0x00}...)
-
-	//The resulting value is then AES-CM- encrypted using the master key to get the cipher key.
-	block, err := aes.NewCipher(masterKey)
-	if err != nil {
-		return nil, err
-	}
-
-	block.Encrypt(sessionKey, sessionKey)
-	return sessionKey, nil
-}
+	out := make([]byte, len(masterSalt))
+	copy(out, masterSalt)
 
-func generateSessionSalt(label byte, masterKey, masterSalt []byte) ([]byte, error) {
-	// https://tools.ietf.org/html/rfc3711#appendix-B.3
-	// The input block for AES-CM is generated by exclusive-oring the master salt with
-	// the concatenation of the encryption salt label
-	sessionSalt := make([]byte, len(masterSalt))
-	copy(sessionSalt, masterSalt)
+	out[7] ^= label
 
-	labelAndIndexOverKdr := []byte{label, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
-	for i, j := len(labelAndIndexOverKdr)-1, len(sessionSalt)-1; i >= 0; i, j = i-1, j-1 {
-		sessionSalt[j] = sessionSalt[j] ^ labelAndIndexOverKdr[i]
-	}
+	// key_id and master_salt are aligned so that their least significant bits agree (right- alignment).
+	lengthDifference := len(masterKey) - len(out)
+	out = append(out, make([]byte, lengthDifference)...)
 
-	// That value is padded and encrypted as above.
-	sessionSalt = append(sessionSalt, []byte{0x00, 0x00}...)
+	//The resulting value is then AES encrypted using the master key to get the cipher key.
 	block, err := aes.NewCipher(masterKey)
 	if err != nil {
 		return nil, err
 	}
 
-	block.Encrypt(sessionSalt, sessionSalt)
-	return sessionSalt[0:len(masterSalt)], nil
+	block.Encrypt(out, out)
+	return out[:outLen], nil
 }
```

Comment:
```go
func aesCmKeyDerivation(label byte, masterKey, masterSalt []byte, indexOverKdr int, outLen int) ([]byte, error) {
  // https://tools.ietf.org/html/rfc3711#appendix-B.3
  // The input block for AES-CM is generated by exclusive-oring the master salt with the
  // concatenation of the encryption key label 0x00 with (index DIV kdr),
  // - index is 'rollover count' and DIV is 'divided by'
  
  nMasterKey := len(masterKey)
  nMasterSalt := len(masterSalt)

  prfIn := make([]byte, nMasterKey)
  copy(prfIn[:nMasterSalt], masterSalt)

  prfIn[7] ^= label
  if indexOverKdr != 0 {
    // 24-bit "index DIV kdr" must be xored to prf input.
    return nil, errors.New("indexOverKdr > 0 is not supported yet")
  }

  //The resulting value is then AES encrypted using the master key to get the cipher key.
  block, err := aes.NewCipher(masterKey)
  if err != nil {
    return nil, err
  }

  out := make([]byte, ((outLen+nMasterKey)/nMasterKey)*nMasterKey)
  var i uint16
  for n := 0; n < outLen; n += nMasterKey {
    binary.BigEndian.PutUint16(prfIn[nMasterKey-2:], i)
    block.Encrypt(out[n:n+nMasterKey], prfIn)
    i++
  }
  return out[:outLen], nil
}
```
unifies `generateSessionAuthTag` into `aesCmKeyDerivation`.

Reply:
Wow this is an amazing cleanup! Thanks for digging so deep into this PR!